### PR TITLE
[12.x] Fix `preg_match()` Deprecation Warning 

### PR DESCRIPTION
**Pull request by sajjadhossainshohag**
Fix `preg_match()` deprecation warning  by handling null values in `isValidUrl ` method.

image

### CODE REVIEW

**Comment by AndrewMast** [comment]
I don't know if this is the best. In this case, if anything but a string is passed in, the result will be `true`. Wouldn't passing in `null` into this new version of `isValidUrl` produce a `true`?

**Comment by sajjadhossainshohag** [comment]
Yes, you are right. I fixed it. 

Thanks.

**Comment by realodix** [comment]
https://github.com/laravel/framework/blob/46ac7829eba556031fa5f540f3771d52fa4117a2/src/Illuminate/Routing/UrlGenerator.php#L669-L672

There seems to be a bug somewhere else. `$path` must only be filled with string, not `null` or others.

**Comment by AndrewMast** [comment]
I honestly think this is getting out of scope. I pointed out the issue with passing in anything but a string (my example was a `null`) to show that this fix doesn't really work. I don't think we want to change any of the docblock parameters.

A solution to this could be setting the type of `$path` to be a string (`isValidUrl(string $path) {}`), but I don't think that matches the style of the code. Another solution could be to change the tests from being negative (`! preg_match`, `!== false`) to be positive and have the default return value to be `false` instead of the current `true`.

**Comment by sajjadhossainshohag** [comment]
Thanks. I'm just as confused. I think the third option would be better. What's the best way to proceed?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
This change is not correct IMO. The bug is that `isValidUrl` is called with something that is not a string. The correct fix is to make sure the caller does the null check.

File: src/Illuminate/Routing/UrlGenerator.php [hunk 1]
```diff
@@ -671,7 +671,7 @@ public function format($root, $path, $route = null)
      */
     public function isValidUrl($path)
     {
-        if (! preg_match('~^(#|//|https?://|(mailto|tel|sms):)~', $path)) {
+        if (! preg_match('~^(#|//|https?://|(mailto|tel|sms):)~', (string) $path)) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/UrlGenerator.php:674]: This is also the wrong place to put this. It needs to go in the **caller** of `isValidUrl` not inside the method body. The place that calls `isValidUrl` needs to make sure to only pass a string.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by GrahamCampbell** [comment]
Oh, the bug lies in your code. `asset(null)` not valid. You cannot pass null as param 1 to the asset function. You MUST pass a string.

**Comment by AndrewMast** [comment]
@GrahamCampbell I think the main issue here is we aren’t actually _enforcing_ a type, we are saying it should be a string through the docblock. Since you are saying that we must enforce it on each call, can we set the parameter type to `string $path`?
